src/android/Dgram.java

```diff
@@ -1,69 +1,42 @@
 package org.apache.cordova.dgram;
 
 import android.util.Log;
-import android.util.SparseArray;
 
 import org.apache.cordova.CallbackContext;
 import org.apache.cordova.CordovaPlugin;
+import org.apache.cordova.PluginResult;
 import org.json.JSONArray;
 import org.json.JSONException;
 import org.json.JSONObject;
 
-import java.io.IOException;
-import java.io.PrintWriter;
-import java.io.StringWriter;
 import java.net.DatagramPacket;
 import java.net.DatagramSocket;
 import java.net.InetAddress;
-import java.net.InetSocketAddress;
-import java.net.NetworkInterface;
+
 import java.net.SocketException;
+import java.nio.charset.StandardCharsets;
 
-import java.util.concurrent.ArrayBlockingQueue;
 
 public class Dgram extends CordovaPlugin {
     private static final String TAG = Dgram.class.getSimpleName();
 
-    SparseArray<DatagramSocket> m_datagramSockets;
-    SparseArray<DatagramSocketListener> m_datagramSocketListeners;
-    ArrayBlockingQueue<String> m_javascriptQueue;
-    SendJavascript m_sendJavascript;
-
-    public Dgram() {
-        m_datagramSockets = new SparseArray<DatagramSocket>();
-        m_datagramSocketListeners = new SparseArray<DatagramSocketListener>();
-        m_javascriptQueue = new ArrayBlockingQueue<String>(1000);
-        m_sendJavascript = new SendJavascript();
-        m_sendJavascript.start();
-    }
+    private static final String OPEN_ACTION = "open";
+    private static final String ON_MESSAGE_ACTION = "onMessage";
+    private static final String SEND_ACTION = "send";
+    private static final String CLOSE_ACTION = "close";
 
-    private class SendJavascript extends Thread {
-        public void run() {
-            final String CORDOVA_REQUIRE = "cordova.require('cordova-plugin-dgram.dgram').";
-            final int JAVASCRIPT_SLEEP = 50;
+    private DatagramSocket datagramSocket;
+    private DatagramSocketListener datagramSocketListener;
 
-            while(true) {
-                try {
-                    String javascript = Dgram.this.m_javascriptQueue.take();
-                    Dgram.this.webView.sendJavascript(CORDOVA_REQUIRE + javascript);
-                    Thread.sleep(JAVASCRIPT_SLEEP);
-                } catch (InterruptedException e) {
-                    Log.d(TAG, e.toString());
-                }
-            }
-        }
-    }
+    private CallbackContext onMessageCallback;
+
+    public Dgram() { }
 
     private class DatagramSocketListener extends Thread {
-        int m_datagramSocketId;
-        DatagramSocket m_datagramSocket;
-
-        public DatagramSocketListener(
-                int datagramSocketId,
-                DatagramSocket datagramSocket
-        ) {
-            this.m_datagramSocketId = datagramSocketId;
-            this.m_datagramSocket = datagramSocket;
+        DatagramSocket datagramSocket;
+
+        DatagramSocketListener() {
+            this.datagramSocket = datagramSocket;
```

Comment:
i find this mechanism for assigning the enclosing classes datagramSocket to the thread confusing.  why not pass datagramSocket into the constructor for the DatagramSocketListener?